### PR TITLE
feat: Compile grainc to JS & create binaries with pkg

### PR DESCRIPTION
**Pull request by phated**
All my current work to get GrainJS working.

I believe I have some other ideas to actually just compile `grainc` to JS, so I've marked this as draft.

### CODE REVIEW

**Comment by phated** [comment]
@ospencer @peblair I believe this is pretty much ready from the packaging/virtual file system perspective. We just need to figure out the String allocation bug that happens in the JS version of the compiler. I would be forever grateful if either of you could help me debug it.

I'll probably be splitting this PR into small stuff that is more related to cleanup of the monorepo, but needed for this stuff to wrap up!

**Comment by ospencer** [comment]
Awesome work! 🎉  

I'll take a look at the allocation stuff.

**Comment by hhugo** [comment]
Is there anything to contribute back to js_of_ocaml ? Better support for unix stubs with node ?

**Comment by phated** [comment]
@hhugo yeah! If you think the unix stubs would be helpful to JSOO, I'm happy to contribute them back!

File: cli/bin/compile.js [hunk 1]
```diff
@@ -1,18 +1,14 @@
...
-        options.cflags ? options.cflags : ""
-      } ${file}`
-    );
+    // TODO: Should we console.log the output (stdout) of the exec?
+    exec(file, options);
     return file.replace(/\.gr$/, ".gr.wasm");
   } catch (e) {
     console.log(e.stdout.toString());
+    // TODO: Do we also want to print the stderr?
+    // console.log(e.stderr.toString());
```
Comment (ospencer) [inline cli/bin/compile.js:11]: Yup, let's go ahead and do these.

File: cli/bin/grain.js [hunk 2]
```diff
@@ -42,9 +40,8 @@ function num(val) {
 }
 
 function graincVersion() {
-  const grainc = path.join(__dirname, "grainc.exe");
-
-  return execSync(`${grainc} --version`).toString().trim();
+  // TODO: This prints "unknown" for the JSOO version of grainc
```
Comment (ospencer) [inline cli/bin/grain.js:43]: I think you said you wanted to just get rid of this? Probably fine if we're bundling the compiler in anyway.

File: cli/bin/compile.js [hunk 3]
```diff
@@ -1,18 +1,10 @@
...
   try {
-    execSync(
-      `${grainc} --stdlib=${options.stdlib} ${
-        options.cflags ? options.cflags : ""
-      } ${file}`
-    );
+    exec(file, options, { stdio: "inherit" });
     return file.replace(/\.gr$/, ".gr.wasm");
   } catch (e) {
-    console.log(e.stdout.toString());
```
Comment (phated) [inline cli/bin/compile.js:15]: @ospencer using `stdio: inherit` here instead of explicitly console.logging

File: cli/bin/exec.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+const path = require("path");
+const { execSync } = require("child_process");
+const fs = require("fs");
+
+function getGrainc() {
+  const grainc = path.join(__dirname, "grainc.exe");
+
+  // TODO: Maybe make an installable path & check it?
```
Comment (phated) [inline cli/bin/exec.js:8]: Needs an issue

Comment (phated) [reply cli/bin/exec.js:8]: #589

File: cli/bin/lsp.js [hunk 5]
```diff
@@ -1,23 +1,14 @@
...
-    execSync(
-      `${grainc} --stdlib=${options.stdlib} ${
-        options.cflags ? options.cflags : ""
-      } -g --lsp ${file}`,
-      { stdio: "inherit" }
-    );
+    exec(`-g --lsp ${file}`, options, { stdio: "inherit" });
     process.exit();
   } catch (e) {
-    console.log(e);
```
Comment (phated) [inline cli/bin/lsp.js:20]: @ospencer this uses `stdio: inherit` already, so we don't need to log the err here.

**Comment by phated** [comment]
@ospencer hmm, there seems to be an issue with the lsp mode that I don't quite understand

**Review by ospencer — APPROVED** [review]
Approved with passing tests!

**Pull request merged**